packages/web5/src/dwn-api.ts

```diff
@@ -0,0 +1,169 @@
+}
+
+export type RecordsQueryResponse = MessageReply & {
+  records: Record[]
+};
+
+export type RecordsDeleteRequest = {
+  author: string;
+  message: Omit<RecordsDeleteOptions, 'authorizationSignatureInput'>;
+}
+
+export type RecordsDeleteResponse = MessageReply;
+
+export type RecordsReadRequest = {
+  author: string;
+  message: Omit<RecordsReadOptions, 'authorizationSignatureInput'>;
+}
+
+export type RecordsReadResponse = MessageReply;
+
+export type ProtocolsConfigureRequest = {
+  author: string;
+  message: Omit<ProtocolsConfigureOptions, 'authorizationSignatureInput'>;
+}
+
+export type ProtocolsQueryRequest = {
+  author: string;
+  message: Omit<ProtocolsQueryOptions, 'authorizationSignatureInput'>
+}
+
+export const DwnApi = (web5Agent: Web5Agent) => ({
+  protocols: {
+    async configure(target: string, request: ProtocolsConfigureRequest) {
+      return await web5Agent.processDwnRequest({
+        target         : target,
+        author         : request.author,
+        messageOptions : request.message,
+        messageType    : DwnInterfaceName.Protocols + DwnMethodName.Configure
+      });
+    },
+    async query(target: string, request: ProtocolsQueryRequest) {
+      return await web5Agent.processDwnRequest({
+        target         : target,
+        author         : request.author,
+        messageOptions : request.message,
+        messageType    : DwnInterfaceName.Protocols + DwnMethodName.Query
+      });
+    }
+  },
+  records: {
+    async write(target: string, request: RecordsWriteRequest): Promise<RecordsWriteResponse> {
+      const { author, data, message: requestMessage } = request;
+      const messageOptions: Partial<RecordsWriteOptions> = {
+        ...requestMessage
+      };
+
+      let dataStream: _Readable.Readable;
+
+      if (data instanceof Blob || data instanceof ReadableStream) {
+        //! TODO: get dataSize and dataCid of data
+      } else {
+        const { dataBytes, dataFormat } = dataToBytes(request.data, messageOptions.dataFormat);
+        messageOptions.data = dataBytes;
+        messageOptions.dataFormat = dataFormat;
+        dataStream = DataStream.fromBytes(dataBytes);
+      }
+
+      const resp = await web5Agent.processDwnRequest({
+        author,
+        dataStream,
+        messageType: DwnInterfaceName.Records + DwnMethodName.Write,
+        messageOptions,
+        target
+      });
+
+      const { message: responseMessage, reply } = resp;
+
+      if (reply.status.code >= 400) {
+        throw new Error('TODO: figure out whether we are throwing an exception');
+      }
+
+      const recordOptions = {
+        author,
+        target,
+        ...responseMessage as RecordsWriteMessage,
+      };
+
+      const record = new Record(web5Agent, recordOptions);
+
+      return { ...reply, record };
+    },
+
+    async create(_target: string, _request: RecordsWriteRequest): Promise<RecordsWriteResponse> {
+      // TODO: damn i guess we have to turn this into a class.
```

Comment:
Why? Is there some sort of encapsulation requirement for creating records? Such as the need for instance properties or methods?

Reply:
great q. `create` is supposed to be an alias for `write`. in order for me to call `write` from within `create` i'd have to call the `DwnApi` function again